### DISCUSSION TITLE
如何二次封装ConfigProvider并做成组件库，让子组件消费统一的样式

### DISCUSSION THREAD

**Discussion by GuanJdoJ** [ant-design/ant-design#43331] [Q&A] [answered]
我想二次封装ConfigProvider，设置一些默认的主题样式，能够让只写一个封装后的Provider就可以消费样式，比如

1. 先提供一个拥有统一样式的Provider 
```tsx
import { ConfigProvider } from 'antd';
import * as React from 'react';

// my-theme-provider.tsx 
const ThemeProvider: React.FC<{
  children?: React.ReactNode;
}> = (props) => {
  return (
      <ConfigProvider
        theme={{
          token: {
            colorPrimary: '#008363',
          },
        }}
      >
        {props.children}
      </ConfigProvider>
  );
};

export default ThemeProvider;
```

2. 把根节点放到ThemeProvider下，消费样式
```tsx
import {ThemeProvider} from 'my-theme-provider-library';
import {Button} from "antd";

const App = () => {
  return <div className='container'>
    <Button type='primary'>按钮</Button>
  </div>
}

export default () => {
  return (
    <ThemeProvider>
      <App />
    </ThemeProvider>
  )
}
```

在同一个项目中这样使用是没有问题的，但是当我把ThemeProvider用dumi做成组件库(my-theme-provider-library)发布后，子组件并没有被设置为统一的样式，这是为什么呢？

**Comment by RedJue** [comment, marked as answer]
看下 antd 的版本是否一致，因为 cp 是消费 antd 上下文的，如果版本不一样就会导致消费的也不一样，就没法生效。

**Reply by GuanJdoJ** [reply]
是的，组件库的和消费项目使用的版本不一致，当我把两个版本设为一样的，样式就能统一了；

但是这样会有一个致命的问题，无法保证消费项目的版本与组件库一致，我能想到的办法是在打包时exclude antd，并且配置antd在peerDependencies，这样才能统一使用消费项目的antd；

还有其他的思路能实现我这个需求吗

**Reply by RedJue** [reply]
最优解应该还是加 peerDep，并且组件库将 antd 作为开发依赖，这样用的时候能保证 antd 是版本唯一的。

**Reply by GuanJdoJ** [reply]
好的，谢谢😘

**Comment by arvinxx** [comment]
组件库里的 antd 记得要设置为 peerDep

**Reply by GuanJdoJ** [reply]
嗯嗯，我能想到的也是这种解决办法，但是我有几个问题想请教下：

1. 设置peerDependencies后，如果不做特殊的处理，antd会被打到输出包中吗；

2. 当消费项目中是固定的antd@5.6.3，但组件库的peerDependencies中是antd: "^5.6.4"，那么1的答案如果是“是”，那么antd的版本是否又会不一致从而安装两个依赖？如果1的答案是“否”，那么消费项目安装时除了warning，还会有其他问题吗？

**Reply by arvinxx** [reply]
> 1. 设置peerDependencies后，如果不做特殊的处理，antd会被打到输出包中吗；

不会

> 当消费项目中是固定的antd@5.6.3，但组件库的peerDependencies中是antd: "^5.6.4"，那么1的答案如果是“是”，那么antd的版本是否又会不一致从而安装两个依赖？如果1的答案是“否”，那么消费项目安装时除了warning，还会有其他问题吗？

1. 如果你的组件库设置了 peerDependencies `antd: "^5.6.4"` 而项目里 固定 `antd:5.6.3`，理论上安装时就会报 peer 不匹配的错误；
2. 建议使用 `>=` 这样的标记，因为如果未来升级到 antd 6， `antd: "^5.6.4"` 这样的写法也会有 peer 不匹配的报错，哪怕实际上可能是兼容的。

**Reply by arvinxx** [reply]
消费项目安装时，可能再关注下 monorepo的场景。单repo基本上不会有问题的。monorepo 视情况可能会有问题。

**Reply by GuanJdoJ** [reply]
好的，谢谢
